File: src/web_app/modern_web/legacy_desktop/objects/prop/prop_rot_angle_manager.py
from __future__ import annotations
from typing import TYPE_CHECKING

from data.constants import (
    BOX,
    CLOCK,
    COUNTER,
    DIAMOND,
    EAST,
    IN,
    NORTH,
    NORTHEAST,
    NORTHWEST,
    OUT,
    SOUTH,
    SOUTHEAST,
    SOUTHWEST,
    WEST,
)

if TYPE_CHECKING:
    from objects.prop.prop import Prop


class PropRotAngleManager:
    def __init__(self, prop: "Prop") -> None:
        self.prop = prop

    def get_diamond_rotation_angle(self) -> int:
        angle_map = {
            IN: {NORTH: 90, SOUTH: 270, WEST: 0, EAST: 180},
            OUT: {NORTH: 270, SOUTH: 90, WEST: 180, EAST: 0},
            CLOCK: {NORTH: 0, SOUTH: 180, WEST: 270, EAST: 90},
            COUNTER: {NORTH: 180, SOUTH: 0, WEST: 90, EAST: 270},
        }

        # Get orientation with fallback to IN
        key = (
            self.prop.state.ori
            if hasattr(self.prop, "state") and hasattr(self.prop.state, "ori")
            else IN
        )
        # Get location with fallback to NORTH
        loc = (
            self.prop.state.loc
            if hasattr(self.prop, "state")
            and hasattr(self.prop.state, "loc")
            and self.prop.state.loc
            else NORTH
        )
        rotation_angle = angle_map.get(key, {}).get(loc, 0)
        return (
            rotation_angle
            if not hasattr(self.prop, "prop_type_str")
            or self.prop.prop_type_str != "Hand"
            else 0
        )

    def get_box_rotation_angle(self) -> int:
        angle_map = {
            IN: {NORTHEAST: 135, NORTHWEST: 45, SOUTHWEST: 315, SOUTHEAST: 225},
            OUT: {NORTHEAST: 315, NORTHWEST: 225, SOUTHWEST: 135, SOUTHEAST: 45},
            CLOCK: {NORTHEAST: 45, NORTHWEST: 315, SOUTHWEST: 225, SOUTHEAST: 135},
            COUNTER: {NORTHEAST: 225, NORTHWEST: 135, SOUTHWEST: 45, SOUTHEAST: 315},
        }
        # Get orientation with fallback to IN
        key = (
            self.prop.state.ori
            if hasattr(self.prop, "state") and hasattr(self.prop.state, "ori")
            else IN
        )
        # Get location with fallback to NORTHEAST
        loc = (
            self.prop.state.loc
            if hasattr(self.prop, "state")
            and hasattr(self.prop.state, "loc")
            and self.prop.state.loc
            else NORTHEAST
        )
        rotation_angle = angle_map.get(key, {}).get(loc, 0)
        return (
            rotation_angle
            if not hasattr(self.prop, "prop_type_str")
            or self.prop.prop_type_str != "Hand"
            else 0
        )

    def update_prop_rot_angle(self) -> None:
        # Default to DIAMOND grid mode
        grid_mode = DIAMOND

        # Determine grid mode based on location if loc exists
        if (
            hasattr(self.prop, "state")
            and hasattr(self.prop.state, "loc")
            and self.prop.state.loc
        ):
            if self.prop.state.loc in ["n", "e", "s", "w"]:
                grid_mode = DIAMOND
            elif self.prop.state.loc in ["ne", "nw", "se", "sw"]:
                grid_mode = BOX

        # Default rotation angle
        prop_rotation_angle = 0

        # Calculate rotation angle based on grid mode
        if grid_mode == DIAMOND:
            prop_rotation_angle = self.get_diamond_rotation_angle()
        elif grid_mode == BOX:
            prop_rotation_angle = self.get_box_rotation_angle()

        # Apply rotation
        self.prop.setTransformOriginPoint(self.prop.boundingRect().center())
        self.prop.setRotation(prop_rotation_angle)

File: web/src/lib/services/implementations/PropRotAngleManager.ts
import { GridMode, Location, Orientation } from "$domain";

/**
 * Calculates prop rotation angles based on location, orientation, and grid mode.
 * Uses static lookup tables for optimal performance.
 */
export class PropRotAngleManager {
  private readonly location: Location;
  private readonly orientation: Orientation;
  private readonly gridMode: GridMode;

  /** Diamond grid angle mappings */
  private static readonly DIAMOND_ANGLE_MAP: Record<
    Orientation,
    Record<Location, number>
  > = {
    [Orientation.IN]: {
      [Location.NORTH]: 90,
      [Location.SOUTH]: 270,
      [Location.WEST]: 0,
      [Location.EAST]: 180,
    } as Record<Location, number>,
    [Orientation.OUT]: {
      [Location.NORTH]: 270,
      [Location.SOUTH]: 90,
      [Location.WEST]: 180,
      [Location.EAST]: 0,
    } as Record<Location, number>,
    [Orientation.CLOCK]: {
      [Location.NORTH]: 0,
      [Location.SOUTH]: 180,
      [Location.WEST]: 270,
      [Location.EAST]: 90,
    } as Record<Location, number>,
    [Orientation.COUNTER]: {
      [Location.NORTH]: 180,
      [Location.SOUTH]: 0,
      [Location.WEST]: 90,
      [Location.EAST]: 270,
    } as Record<Location, number>,
  };

  /** Box grid angle mappings */
  private static readonly BOX_ANGLE_MAP: Record<
    Orientation,
    Record<Location, number>
  > = {
    [Orientation.IN]: {
      [Location.NORTHEAST]: 135,
      [Location.NORTHWEST]: 45,
      [Location.SOUTHWEST]: 315,
      [Location.SOUTHEAST]: 225,
    } as Record<Location, number>,
    [Orientation.OUT]: {
      [Location.NORTHEAST]: 315,
      [Location.NORTHWEST]: 225,
      [Location.SOUTHWEST]: 135,
      [Location.SOUTHEAST]: 45,
    } as Record<Location, number>,
    [Orientation.CLOCK]: {
      [Location.NORTHEAST]: 45,
      [Location.NORTHWEST]: 315,
      [Location.SOUTHWEST]: 225,
      [Location.SOUTHEAST]: 135,
    } as Record<Location, number>,
    [Orientation.COUNTER]: {
      [Location.NORTHEAST]: 225,
      [Location.NORTHWEST]: 135,
      [Location.SOUTHWEST]: 45,
      [Location.SOUTHEAST]: 315,
    } as Record<Location, number>,
  };

  /** Creates a new PropRotAngleManager instance */
  constructor({
    location,
    orientation,
    gridMode,
  }: {
    location: Location;
    orientation: Orientation;
    gridMode: GridMode;
  }) {
    this.location = location;
    this.orientation = orientation;
    this.gridMode = gridMode;
  }

  /** Calculates the rotation angle in degrees */
  getRotationAngle(): number {
    const angleMap =
      this.gridMode === GridMode.DIAMOND
        ? PropRotAngleManager.DIAMOND_ANGLE_MAP
        : PropRotAngleManager.BOX_ANGLE_MAP;

    const orientationAngles = angleMap[this.orientation];
    return orientationAngles?.[this.location] ?? 0;
  }

  /** Static method for rotation angle calculation (recommended) */
  static calculateRotation(
    location: Location,
    orientation: Orientation,
    gridMode: GridMode
  ): number {
    const angleMap =
      gridMode === GridMode.DIAMOND
        ? PropRotAngleManager.DIAMOND_ANGLE_MAP
        : PropRotAngleManager.BOX_ANGLE_MAP;

    const orientationAngles = angleMap[orientation];
    return orientationAngles?.[location] ?? 0;
  }
}

export default PropRotAngleManager;
